### ISSUE TITLE
Support custom evaluation metrics in R

### ISSUE THREAD

**Issue by CliffordLai** [catboost/catboost#350]
Problem:
I found documentation for defining a custom metric function in python from:[https://tech.yandex.com/catboost/doc/dg/concepts/python-usages-examples-docpage/#custom-loss-function-eval-metric](url)
But is there a way to do this in R and how should the function be defined?

catboost version: 0.8
Operating System: Windows 10
CPU: i5-2500k

**Comment by hugoncosta** [comment]
catboost.train(learn_pool, 
               test_pool = NULL, 
               params = list())

In the params you have to define eval_metric = metric_name:param=Value (check the list of parameters [here](https://tech.yandex.com/catboost/doc/dg/concepts/r-reference_catboost-train-docpage/))

**Comment by CliffordLai** [comment]
Thanks, hugoncosta. I found the "eval_metric" and the parameter "custom_loss", which states that

"Metric values to output during training. These functions are not optimized and are displayed for informational purposes only."

However, I don't know how the custom function is defined in R? What is the required format? For example, if I want to define my own RMSE, how should I do it in R for catboost to work?

**Comment by hugoncosta** [comment]
Have you tried the way presented on the [Python tutorial](https://github.com/catboost/catboost/blob/master/catboost/tutorials/catboost_python_tutorial.ipynb)? Basically, defining the function and then custom_loss = my_function().

**Comment by CliffordLai** [comment]
Thank you for the quick reply. I tried defining a function similar to what is presented in Section 3.6 of the python tutorial, but I am not sure how to translate the format from python to R.

Here is my example user-defined RMSE in R. I tried to follow the format of the one given in the python tutorial. Unfortunately, it did not work. 

```
MyRMSE <- function(object) {
  get_final_error <- function(error, weight){
    return(error/(weight+1e-38))
  }

  is_max_optimal <- function(){
    return(FALSE)
  }

  evaluate <- function(approxes, target, weight){
    if(is.null(weight)){
      w <- rep(1,length(target))
    }
    approx <- approxes[[1]]
    error_sum <- sqrt(mean(w*(approx-target)^2))
    weight_sum <- sum(w)
    return(c(error_sum, weight_sum))
  }

  return(list(get_final_error = get_final_error, 
              is_max_optimal = is_max_optimal, 
              evaluate = evaluate))
}
```


I took the example from the R Usage examples of catboost, [https://tech.yandex.com/catboost/doc/dg/concepts/r-usages-examples-docpage/](https://tech.yandex.com/catboost/doc/dg/concepts/r-usages-examples-docpage/).

Try using the following parameter set-up,
```
fit_params <- list(iterations = 20,
                   loss_function = 'Logloss:Border=0.5',
                   custom_loss=MyRMSE,
                   ignored_features = c(4,9),
                   border_count = 32,
                   depth = 5,
                   learning_rate = 0.03,
                   l2_leaf_reg = 3.5,
                   train_dir = 'train_dir',
                   random_seed=111,
                   metric_period=5)
```

It then returns an error:
```
Error in catboost.train(train_pool, test_pool, fit_params) : 
  atboost/catboost/libs/options/loss_description.h:100: function (object)  loss is not supported
```

It seems that the problem is caused by the definition of the function MyRMSE, any idea?

**Comment by hugoncosta** [comment]
That error does seem like CatBoost doesn't support custom functions on R indeed :( Let's see if the developers can check if this is indeed the case.

**Comment by annaveronika** [comment]
Custom metrics are not jet supported for R, but you can write your metric in C++, recompile the package and use it in R, and it's very simple, the tutorial for that is here:
https://github.com/catboost/catboost/blob/master/catboost/tutorials/advanced_tutorials/catboost_custom_metric_tutorial.md

**Comment by CliffordLai** [comment]
That is great! Thank you for pointing it out.

**Comment by elfrasco** [comment]
It seems that link shared by @annaveronika is broken :-(

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
Please try this one https://github.com/catboost/tutorials/blob/master/custom_loss/custom_metric_tutorial.md
